### PR TITLE
Document tedge_actors crate

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

Documentation improvements [WIP]:
* [x] overall crate documentation `lib.rs`
* [x] Actor
* [x] Message Boxes
* [x] Builder
* [ ] Runtime
* [ ] Internal
* [x] Test helpers

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_actors/src/actors.rs [hunk 1]
```diff
@@ -37,6 +37,91 @@ impl<S: Server> ServerActor<S> {
...
+/// #
+/// # #[tokio::main]
+/// # async fn main_test() {
+/// #
+/// // Create an actor to handle the requests to a server
+/// let server = Calculator::default();
+/// let actor = ServerActor::new(server);
+///
+/// // As for any actor, one needs a bidirectional channel to the message box of the server.
+/// let (mut test_box, actor_box) = SimpleMessageBox::channel("Test", 10);
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/actors.rs:104]:
```suggestion
/// let (mut client_box, server_box) = SimpleMessageBox::channel("Test", 10);
```
Just to highlight our server/client notion more explicitly.

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 2]
```diff
@@ -1,2 +1,102 @@
-//! Examples
+//! Let's start be implementing a client actor for the calculator service.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:1]:
```suggestion
//! Let's start be implementing a client actor for the calculator server actor.
```

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 3]
```diff
@@ -1,2 +1,102 @@
...
+//! struct Player {
+//!     name: String,
+//!     target: i64,
+//! }
+//!
+//! #[async_trait]
+//! impl Actor for Player {
+//!
+//!     /// This actor use a simple message box
+//!     /// to receive `Update` messages and to send `Operation` messages.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:18]:
```suggestion
//!     /// to send `Operation` requests and to receive `Update` responses
```
Sounds more natural

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 4]
```diff
@@ -1,2 +1,102 @@
...
+//!             let delta = self.target - status.to;
+//!             messages.send(Operation::Add(delta / 2)).await?;
+//!         }
+//!
+//!         Ok(())
+//!     }
+//! }
+//! ```
+//!
+//! To connect such an actor to the calculator, one needs message box builders
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:44]:
```suggestion
//! To connect such an actor to the calculator, one needs actor builders
```

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 5]
```diff
@@ -1,2 +1,102 @@
...
+//! ```
+//! # use tedge_actors::{Actor, Builder, ChannelError, MessageBox, MessageBoxPlug, NoConfig, ServiceActor, ServiceMessageBox, ServiceMessageBoxBuilder, SimpleMessageBox, SimpleMessageBoxBuilder};
+//! # use crate::tedge_actors::examples::calculator::*;
+//! # #[tokio::main]
+//! # async fn main_test() -> Result<(),ChannelError> {
+//! #
+//!
+//! // Building a box to hold 16 pending requests for the calculator service
+//! // Note that a service actor requires a specific type of message box.
+//! let mut service_box_builder = ServiceMessageBoxBuilder::new("Calculator", 16);
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:56]:
```suggestion
//! let mut service_box_builder = ServerMessageBoxBuilder::new("Calculator", 16);
```

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 6]
```diff
@@ -1,2 +1,102 @@
...
+//! to establish appropriate connections between the actor message boxes.
+//!
+//! ```
+//! # use tedge_actors::{Actor, Builder, ChannelError, MessageBox, MessageBoxPlug, NoConfig, ServiceActor, ServiceMessageBox, ServiceMessageBoxBuilder, SimpleMessageBox, SimpleMessageBoxBuilder};
+//! # use crate::tedge_actors::examples::calculator::*;
+//! # #[tokio::main]
+//! # async fn main_test() -> Result<(),ChannelError> {
+//! #
+//!
+//! // Building a box to hold 16 pending requests for the calculator service
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:54]:
```suggestion
//! // Building a box to hold 16 pending requests for the calculator server
```

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 7]
```diff
@@ -1,2 +1,102 @@
...
+//! // Note that a service actor requires a specific type of message box.
+//! let mut service_box_builder = ServiceMessageBoxBuilder::new("Calculator", 16);
+//!
+//! // Building a box to hold one pending message for the player
+//! // This actor never expect more then one message.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:59]:
```suggestion
//! // This actor never expects more than one message.
```

File: crates/core/tedge_actors/src/examples/mod.rs [hunk 8]
```diff
@@ -1,2 +1,102 @@
...
+//! player_2_box_builder.connect_to(&mut service_box_builder, NoConfig);
+//!
+//! // One can then build the message boxes
+//! let service_box: ServiceMessageBox<Operation,Update> = service_box_builder.build();
+//! let mut player_1_box = player_1_box_builder.build();
+//! let mut player_2_box = player_2_box_builder.build();
+//!
+//! // Then spawn the service
+//! let service = Calculator::default();
+//! tokio::spawn(ServiceActor::new(service).run(service_box));
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/examples/mod.rs:79]:
```suggestion
//! tokio::spawn(ServerActor::new(service).run(service_box));
```

File: crates/core/tedge_actors/src/lib.rs [hunk 9]
```diff
@@ -110,260 +123,54 @@
...
-//! # }
+//! They interact by sending messages and a key step is to connect the actors with each other.
+//!
+//! Connection between actors are established using [actor and message box builders](crate::builders).
+//! These builders implement connector traits
+//! that define the services provided and consumed by the actors under construction.
+//!
+//! The connection builder traits work as pairs:
+//! - An actor that provides some service makes this service available
+//!   with a connection builder that implements the [ServiceProvider](crate::ServiceProvider) trait.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/lib.rs:143]:
```suggestion
//!   with an actor builder that implements the [ServiceProvider](crate::ServiceProvider) trait.
```

File: crates/core/tedge_actors/src/lib.rs [hunk 10]
```diff
@@ -110,260 +123,54 @@
...
+//! - In a symmetrical way, an actor that requires an other service to provide its own feature,
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/lib.rs:144]:
```suggestion
//! - In a symmetrical way, an actor that requires another service to provide its own feature,
```

File: crates/core/tedge_actors/src/lib.rs [hunk 11]
```diff
@@ -110,260 +123,54 @@
...
+//!   makes the connection feasible with a connection builder
+//!   that implements the [ServiceConsumer](crate::ServiceConsumer) trait.
+//! - These two traits define the types of the messages sent in both directions
+//!   and how to connect the message boxes of the actors under construction,
+//!   possibly using some configuration.
+//! - Two actor builders, a `consumer: ServiceConsumer<I,O>` and a `producer: ServiceProducer<I,O,C>`
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/lib.rs:150]:
```suggestion
//! - Two actor builders, a `consumer: ServiceConsumer<I,O>` and a `producer: ServiceProvider<I,O,C>`
```

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 12]
```diff
@@ -16,6 +16,61 @@ use std::fmt::Debug;
...
+///
+/// ```
+/// # use tedge_actors::{Actor, Builder, ChannelError, ServiceConsumer, NoConfig, ServerActor, ServerMessageBoxBuilder, SimpleMessageBoxBuilder};
+///
+/// # use tedge_actors::test_helpers::ProbeEvent::{Recv, Send};
+/// # use crate::tedge_actors::examples::calculator::*;
+/// # #[tokio::main]
+/// # async fn main_test() -> Result<(),ChannelError> {
+/// #
+/// // The `MessageBoxPlugExt` must be visible to add the `with_probe()` method on actor builders
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/test_helpers.rs:38]:
```suggestion
/// // The `ServiceConsumerExt` must be in scope to use the `with_probe()` method on actor builders
```

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 13]
```diff
@@ -123,7 +202,31 @@ impl<I: MessagePlus, O: MessagePlus> Probe<I, O> {
     }
 }
 
+/// Extend any `MessageBoxPlug` with a `with_probe` method.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/test_helpers.rs:205]:
```suggestion
/// Extend any `ServiceConsumerExt` with a `with_probe` method.
```

**Review by albinsuresh — COMMENTED** [review]
Observed a few recurring issues:
1. There are still references to the old service concept which was renamed as server recently
2. The terms actor builder and message box builders are still used inconsistently. We probably need to stick to one (probably the former)

File: crates/core/tedge_actors/src/test_helpers.rs (hunk 13 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/test_helpers.rs]: Fixed

File: crates/core/tedge_actors/src/test_helpers.rs (hunk 12 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/test_helpers.rs]: fixed

File: crates/core/tedge_actors/src/lib.rs (hunk 11 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/lib.rs]: Fixed

File: crates/core/tedge_actors/src/lib.rs (hunk 10 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/lib.rs]: fixed

File: crates/core/tedge_actors/src/lib.rs (hunk 9 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/lib.rs]: fixed

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 8 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: Fixed

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 7 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: fixed

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 6 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: Fixed

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 5 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: Fixed

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 4 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: done

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: done

File: crates/core/tedge_actors/src/examples/mod.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/examples/mod.rs]: done

File: crates/core/tedge_actors/src/actors.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/actors.rs]: Done

**Comment by didier-wenzek** [comment]
> 1. There are still references to the old service concept which was renamed as server recently(I hope I have spotted and highlighted them all)

Yes, it seems I messed up things while rebasing. Fixed.

> 2. The terms actor builder and message box builders are still used inconsistently. We probably need to stick to one (probably the former)

Agree. I will stick on the *actor builder* term.

File: crates/core/tedge_actors/src/lib.rs [hunk 14]
```diff
@@ -111,263 +124,76 @@
...
+//! #     }
 //! # }
+//!
+//! // An actor builder defines the service provided by that actor
+//! // by implementing the `ServiceProvider` trait for the appropriate input, output and config types.
+//! //
+//! // Here, `SomeActorBuilder: ServiceProvider<SomeInput, SomeOutput, SomeConfig>`
+//! let mut producer = SomeActorBuilder::default();
+//!
+//! // An actor builder also defines the services required by the actor. This is done
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/lib.rs:185]:
```suggestion
//! // An actor builder also declares that it is a consumer of other services required by it. This is done
```

File: crates/core/tedge_actors/src/lib.rs [hunk 15]
```diff
@@ -111,263 +124,76 @@
...
+//! // by implementing the `ServiceConsumer` trait for the appropriate input, output and config types.
+//! //
+//! // Here, `SomeOtherActorBuilder: ServiceConsumer<SomeInput, SomeOutput>`
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/lib.rs:188]:
```suggestion
//! // Here, `SomeOtherActorBuilder: ServiceConsumer<SomeInput, SomeOutput, SomeConfig>`
```

File: crates/core/tedge_actors/src/message_boxes.rs [hunk 16]
```diff
@@ -82,10 +82,10 @@
 //!
 //! This crates provides several built-in message box implementations:
 //!
-//! - [SimpleMessageBox<I,O>](crate::SimpleMessageBox) for actors that simply process messages in turn,
-//! - [ServerMessageBox<I,O>](crate::ServerMessageBox) for actors that deliver a request-response server,
-//! - [ConcurrentServerMessageBox<I,O>](crate::ConcurrentServerMessageBox) for server actors that process requests concurrently,
-//! - [ClientMessageBox<I,O>](crate::ClientMessageBox) for actors that use a request-response service,
+//! - [SimpleMessageBox](crate::SimpleMessageBox) for actors that simply process messages in turn,
+//! - [ServerMessageBox](crate::ServerMessageBox) for actors that deliver a request-response service,
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/message_boxes.rs:86]:
```suggestion
//! - [ServerMessageBox](crate::ServerMessageBox) for server actors that deliver a request-response service,
```

File: crates/core/tedge_actors/src/message_boxes.rs [hunk 17]
```diff
@@ -82,10 +82,10 @@
...
+//! - [ConcurrentServerMessageBox](crate::ConcurrentServerMessageBox) for service actors that process requests concurrently,
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/message_boxes.rs:87]:
```suggestion
//! - [ConcurrentServerMessageBox](crate::ConcurrentServerMessageBox) for server actors that process requests concurrently,
```

File: crates/core/tedge_actors/src/message_boxes.rs [hunk 18]
```diff
@@ -82,10 +82,10 @@
...
+//! - [ClientMessageBox](crate::ClientMessageBox) for actors that use a request-response service,
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/message_boxes.rs:88]:
```suggestion
//! - [ClientMessageBox](crate::ClientMessageBox) for client actors that use a request-response service from a server actor,
```

File: crates/core/tedge_actors/src/message_boxes.rs [hunk 19]
```diff
@@ -288,6 +288,7 @@ impl<Input: Message, Output: Message> MessageBox for SimpleMessageBox<Input, Out
 pub type ServerMessageBox<Request, Response> =
     SimpleMessageBox<(ClientId, Request), (ClientId, Response)>;
 
+/// Internal id assigned to a client actor of service actor
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/message_boxes.rs:291]:
```suggestion
/// Internal id assigned to a client actor of a server actor
```

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 20]
```diff
@@ -65,6 +138,12 @@ impl<I: MessagePlus, O: MessagePlus> Probe<I, O> {
         }
     }
 
+    /// Return the next event observed between the two connected actors.
+    ///
+    /// Note that calling `observe()` is mandatory for the actors to make progress.
+    /// Indeed the observed channels are blocked on each action till the event is actually observed.
+    /// Hence a probe also control at which pace the input/output messages are sent
+    /// between the two actors which connection is observed.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/test_helpers.rs:146]:
```suggestion
    /// Hence a probe also controls the pace at which input/output messages are sent
    /// over the observed connection between the two actors
```

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 21]
```diff
@@ -127,7 +206,31 @@ impl<I: MessagePlus, O: MessagePlus> Probe<I, O> {
     }
 }
 
+/// Extend any [ServiceConsumer] with a `with_probe` method.
 pub trait ServiceConsumerExt<Request: MessagePlus, Response: MessagePlus> {
+    /// Add a probe to `self` that is a [ServiceConsumer](crate::ServiceConsumer) of a first actor.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/test_helpers.rs:211]:
```suggestion
    /// Add a probe to an actor `self` that is a [ServiceConsumer](crate::ServiceConsumer).
```

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 22]
```diff
@@ -127,7 +206,31 @@ impl<I: MessagePlus, O: MessagePlus> Probe<I, O> {
...
+    ///
+    /// Return a [ServiceConsumer](crate::ServiceConsumer)
+    /// ready to be plug into a [ServiceProvider](crate::ServiceProvider) of a second actor.
```
Comment (albinsuresh) [inline crates/core/tedge_actors/src/test_helpers.rs:214]:
```suggestion
    /// that can be plugged into another actor which is a [ServiceProvider](crate::ServiceProvider).
```

**Review by albinsuresh — APPROVED** [review]

**Pull request merged**